### PR TITLE
Add iptables rules to deny unwanted VNC access (bsc#938200)

### PR DESCRIPTION
**Pull request by dguitarbite**
VNC ports are accessible from the management network. This patch will
allow us to restrict the VNC console from unwanted access to VNC
console. Using iptables u32 and string module to filter out VNC packets.

### CODE REVIEW

**Comment by tboerger** [comment]
Why do you close https://github.com/crowbar/crowbar-openstack/pull/5 and reopen it against tex?

File: chef/cookbooks/nova/recipes/compute.rb [hunk 1]
```diff
@@ -423,3 +429,36 @@ def set_boot_kernel_and_trigger_reboot(flavor='default')
...
+bash "nova_compute_vncblock_reject_all" do
+  code <<-EOH
+    iptables -D INPUT -j VNCBLOCK -p tcp --match multiport --dports 5900:15900
+    iptables -I INPUT -j VNCBLOCK -p tcp --match multiport --dports 5900:15900
+    iptables -F VNCBLOCK
+    iptables -I VNCBLOCK -p tcp \
+        -m connbytes --connbytes 0:1024 \
+        --connbytes-dir both --connbytes-mode bytes \
+        -m state --state ESTABLISHED \
+        -m u32 --u32 "0>>22&0x3C@ 12>>26&0x3C@ 0=0x52464220" \
```
Comment (ovisan) [inline chef/cookbooks/nova/recipes/compute.rb]: Can you please explain what this cryptic line does?

Comment (dguitarbite) [inline chef/cookbooks/nova/recipes/compute.rb]: Snoops all packets from port range 5900:15900 to check for VNC packets and filters them accordingly.

**Comment by dguitarbite** [comment]
@tboerger I will forward port this change.

File: chef/cookbooks/nova/recipes/compute.rb (hunk 1 above)
Comment (bmwiedemann) [inline chef/cookbooks/nova/recipes/compute.rb]: :-1: bad code duplication. the VNC match should go into the -I INPUT line above

Comment (dguitarbite) [inline chef/cookbooks/nova/recipes/compute.rb]: @bmwiedemann I did it on purpose to check for VNC match only for those packets which are under the port range.

Comment (bmwiedemann) [inline chef/cookbooks/nova/recipes/compute.rb]: just add it on the right side above. As it currently is, it is even slower as the match has to be done twice.
apart from being ugly and unmaintable

File: chef/cookbooks/nova/recipes/compute.rb [hunk 2]
```diff
@@ -423,3 +429,36 @@ def set_boot_kernel_and_trigger_reboot(flavor='default')
...
+
+# Set iptables rules for blocking VNC Access for all but the nova-controller node.
+bash "nova_compute_vncblock_reject_all" do
+  code <<-EOH
+    iptables -N VNCBLOCK
+    iptables -I INPUT -j VNCBLOCK -p tcp --match multiport --dports 5900:15900
+    iptables -F VNCBLOCK
+    iptables -I VNCBLOCK -p tcp \
+        -m connbytes --connbytes 0:1024 \
+        --connbytes-dir both --connbytes-mode bytes \
```
Comment (bmwiedemann) [inline chef/cookbooks/nova/recipes/compute.rb]: man iptables-extensions says on this topic

```
NOTE that for connections which have no accounting information, the match will always return false.
```

and cat /proc/sys/net/netfilter/nf_conntrack_acct only shows a 1 for SLE11, but not SLE12 by default, so I guess if we wanted to use this, we would need to enable accounting.

Comment (dguitarbite) [inline chef/cookbooks/nova/recipes/compute.rb]: It shows 1 for me in SLE12.

File: chef/cookbooks/nova/recipes/compute.rb (hunk 1 above)
Comment (dguitarbite) [inline chef/cookbooks/nova/recipes/compute.rb]: Ok, I'll update it. Thanks for catching this.

**Comment by bmwiedemann** [comment]
:+1: iptables rules look good now, except I'm not sure if the nova_controller_ips are the correct ones in case of nova != horizon and in HA case

**Comment by dguitarbite** [comment]
@bmwiedemann As far as I could understand grepping through the barclamps : nova_controller = search(:node, "roles:nova-multi-controller#{env_filter}") should fetch me all the nodes which have nova_multi_controller barclamp installed (which may or may not be dashboard/horizon node), but I am not an expert and failed to find one who could confirm if this is correct. I have done my best here.

**Comment by vuntz** [comment]
Let's not merge this until https://github.com/crowbar/crowbar-openstack/pull/16 is merged.

**Comment by dirkmueller** [comment]
@dguitarbite this needs the changes from the master PR backported.., -1 for now

**Comment by dguitarbite** [comment]
@dirkmueller Backported the changes to master.

**Comment by tboerger** [comment]
LGTM

**Pull request merged**